perllib/FixMyStreet/App/Controller/Admin.pm

```diff
@@ -514,41 +636,165 @@ sub lookup_body : Private {
+        $c->log->debug( 'posted' );
+        $c->forward('update_contacts');
+        $c->forward('update_body_translations');
+    }
+
+    my @translations = $c->model('DB::Translation')->search(
+        {
+            tbl => 'body',
+            object_id => $c->stash->{body_id},
+            col => 'name'
+        }
+    )->all;
+
+    $c->stash->{translations} = \@translations;
+
+    $c->forward('fetch_languages');
+    $c->forward('fetch_contacts');
 
+    $c->stash->{template} = 'admin/body.html';
+    return 1;
+}
+
+sub body_translation_edit : Chained('body') : PathPart('translation') : Args(1) {
+    my ( $self, $c, $lang ) = @_;
+
+    $c->forward('check_for_super_user');
     $c->forward( '/auth/get_csrf_token' );
     $c->forward( 'lookup_body' );
 
+    my $translation = $c->model('DB::Translation')->find(
+        {
+            tbl => 'body',
+            object_id => $c->stash->{body_id},
+            col => 'name',
+            lang => $lang
+        }
+    );
+
+    $c->stash->{translation} = $translation;
+
+    my $language_name = '';
+    foreach my $supported_lang (@{$c->cobrand->languages}) {
+        my ($id, $name, $code) = split(',', $supported_lang);
+        $language_name = $name if $id eq $lang;
+    }
+
+    $c->stash->{language_name} = $language_name;
+
+    return 1;
+}
+
+sub category : Chained('body_base') : PathPart('') : CaptureArgs(2) {
+    my ( $self, $c, $body_id, $category ) = @_;
+
+    $c->forward( '/auth/get_csrf_token' );
+    $c->stash->{body_id} = $body_id;
+    $c->forward( 'lookup_body' );
+
     my $contact = $c->stash->{body}->contacts->search( { category => $category } )->first;
     $c->stash->{contact} = $contact;
+}
+
+sub category_edit : Chained('category') : PathPart('') : Args(0) {
+    my ( $self, $c ) = @_;
+
+    if ( $c->get_param('posted') ) {
+        $c->log->debug( 'posted' );
+        $c->forward('update_category_translations');
+    }
 
     my $history = $c->model('DB::ContactsHistory')->search(
         {
-            body_id => $body_id,
-            category => $category
+            body_id => $c->stash->{body_id},
+            category => $c->stash->{contact}->category
         },
         {
             order_by => ['contacts_history_id']
         },
     );
     $c->stash->{history} = $history;
 
+    my @translations = $c->model('DB::Translation')->search(
+        {
+            tbl => 'contact',
+            object_id => $c->stash->{contact}->id,
+            col => 'category'
+        }
+    )->all;
+
+    $c->forward('fetch_languages');
+    $c->stash->{translations} = \@translations;
+
+
     my @methods = map { $_ =~ s/FixMyStreet::SendReport:://; $_ } keys %{ FixMyStreet::SendReport->get_senders };
     $c->stash->{send_methods} = \@methods;
 
     return 1;
 }
 
+sub category_translation_edit : Chained('category') : PathPart('translation') : Args(1) {
+    my ( $self, $c, $lang ) = @_;
+
+    my $translation = $c->model('DB::Translation')->find(
```

Comment:
Ditto here, could actually be same helper method in fact (if lang/no lang provided).
Perhaps auto-provided by the Role?